api/pull.go

```diff
@@ -0,0 +1,380 @@
+}
+
+// NewBlockFilter creates a filter that fetches blocks that are imported into the chain.
+// It is part of the filter package since polling goes with eth_getFilterChanges.
+func (api *PullAPI) NewBlockFilter() (rpc.ID, error) {
+	last, err := api.blocks.LatestEVMHeight()
+	if err != nil {
+		return "", err
+	}
+
+	f := newBlocksFilter(api.config.FilterExpiry, last)
+	return api.addFilter(f), nil
+}
+
+func (api *PullAPI) UninstallFilter(id rpc.ID) bool {
+	api.mux.Lock()
+	defer api.mux.Unlock()
+
+	_, ok := api.filters[id]
+	delete(api.filters, id)
+	return ok
+}
+
+// NewFilter creates a new filter and returns the filter id. It can be
+// used to retrieve logs when the state changes. This method cannot be
+// used to fetch logs that are already stored in the state.
+//
+// Default criteria for the from and to block are "latest".
+// Using "latest" as block number will return logs for mined blocks.
+// Using "pending" as block number returns logs for not yet mined (pending) blocks.
+// In case logs are removed (chain reorg) previously returned logs are returned
+// again but with the removed property set to true.
+//
+// In case "fromBlock" > "toBlock" an error is returned.
+func (api *PullAPI) NewFilter(criteria filters.FilterCriteria) (rpc.ID, error) {
+	from, err := api.blocks.LatestEVMHeight()
+	if err != nil {
+		return "", err
+	}
+
+	// if from block is actually set we use it as from value otherwise use latest
+	if criteria.FromBlock != nil && criteria.FromBlock.Int64() >= 0 {
+		from = criteria.FromBlock.Uint64()
+		// if to block is set make sure it's not less than from block
+		if criteria.ToBlock != nil && criteria.FromBlock.Cmp(criteria.ToBlock) > 0 {
+			return "", fmt.Errorf("from block must be lower than to block")
+		}
+		// todo we should check for max range of from-to heights
+	}
+
+	f := newLogsFilter(api.config.FilterExpiry, from, &criteria)
+	return api.addFilter(f), nil
+}
+
+// GetFilterLogs returns the logs for the filter with the given id.
+// If the filter could not be found an empty array of logs is returned.
+func (api *PullAPI) GetFilterLogs(ctx context.Context, id rpc.ID) ([]*gethTypes.Log, error) {
+	panic("not implemented")
+}
+
+// GetFilterChanges returns the logs for the filter with the given id since
+// last time it was called. This can be used for polling.
+//
+// For pending transaction and block filters the result is []common.Hash.
+// (pending)Log filters return []Log.
+func (api *PullAPI) GetFilterChanges(id rpc.ID) (interface{}, error) {
+	api.mux.RLock()
+	defer api.mux.RUnlock()
+
+	f, ok := api.filters[id]
+	if !ok {
+		return nil, errors.Join(errs.ErrNotFound, fmt.Errorf("filted by id %s does not exist", id))
+	}
+
+	current, err := api.blocks.LatestEVMHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	if f.expired() {
+		api.UninstallFilter(id)
+		return nil, errors.Join(errs.ErrNotFound, fmt.Errorf("filted by id %s expired", id))
+	}
+	defer f.updateUsed(current)
```

Comment:
Do we want to update this even if the possible calls of:
- `api.getBlocks(current, filter)`
- `api.getTransactions(current, filter)`
- `api.getLogs(current, filter)`

error out? Would it make more sense to update it only when the filter successfully returns a response, and not an error?

Reply:
good point